File: trie/completeString/code.py
class Node:
    def __init__(self):
        self.flag = False
        self.links = [None] * 26
    
    def contains(self, c):
        return self.links[ord(c) - ord('a')] is not None
    
    def put(self, c):
        self.links[ord(c) - ord('a')] = Node()
    
    def next(self, c):
        return self.links[ord(c) - ord('a')]
    
    def setEnd(self):
        self.flag = True
    
    def isEnd(self):
        return self.flag


class Trie:
    def __init__(self):
        self.root = Node()
    
    def insert(self, s):
        tmp = self.root
        for c in s:
            if not tmp.contains(c):
                tmp.put(c)
            tmp = tmp.next(c)
        tmp.setEnd()
    
    def search(self, s):
        tmp = self.root
        for c in s:
            tmp = tmp.next(c)
            if not tmp.isEnd():
                return False
        return True


def completeString(n, a):
    trie = Trie()
    for s in a:
        trie.insert(s)
    
    ans = ""
    for s in a:
        if trie.search(s):
            if len(s) > len(ans) or (len(s) == len(ans) and s < ans):
                ans = s
    
    if len(ans) == 0:
        return "None"
    return ans

File: trie/completeString/code.go
type Node struct {
	flag  bool
	links [26]*Node
}

func (n *Node) contains(c byte) bool {
	return n.links[c-'a'] != nil
}

func (n *Node) put(c byte) {
	n.links[c-'a'] = &Node{}
}

func (n *Node) next(c byte) *Node {
	return n.links[c-'a']
}

func (n *Node) setEnd() {
	n.flag = true
}

func (n *Node) isEnd() bool {
	return n.flag
}

type Trie struct {
	root *Node
}

func NewTrie() *Trie {
	return &Trie{
		root: &Node{},
	}
}

func (t *Trie) insert(s string) {
	tmp := t.root
	for i := 0; i < len(s); i++ {
		c := s[i]
		if !tmp.contains(c) {
			tmp.put(c)
		}
		tmp = tmp.next(c)
	}
	tmp.setEnd()
}

func (t *Trie) search(s string) bool {
	tmp := t.root
	for i := 0; i < len(s); i++ {
		c := s[i]
		tmp = tmp.next(c)
		if !tmp.isEnd() {
			return false
		}
	}
	return true
}

func completeString(n int, a []string) string {
	trie := NewTrie()
	for _, s := range a {
		trie.insert(s)
	}

	var ans string
	for _, s := range a {
		if trie.search(s) {
			if len(s) > len(ans) || (len(s) == len(ans) && s < ans) {
				ans = s
			}
		}
	}

	if len(ans) == 0 {
		return "None"
	}
	return ans
}
